### PR TITLE
make git protocol used to clone, explicit

### PR DESCRIPTION
**Pull request by zancas**
github offers a "clone" url over http..   I believe using `git@github` uses the user's `ssh` keypair to authenticate.   If I'm right, then we should tell users to use that version.

### CODE REVIEW

**Comment by zancas** [comment]
Good point!    I usually clone from the "upstream" repo..   the one I want to contribute to, and then, locally add my fork as something-other-than-origin.

Let me read a bit more so I can better understand the implications of having different git remotes have the alias "origin".

**Comment by zancas** [comment]
It's possible to complete the build without contributing code to the `zcash-gitian` project.

Is instruction about forking necessary for this REAME?

If it is, then I will add instructions for adding and specifying remotes-and-aliases.

I think that the cost is:

  * more verbose README, including not _strictly_ necessary detail.

I think that the benefit is:

  * empowers diligent readers, not already familiar with pull-requests, with a correct first step towards that process.

I think that if we want to support "PR-naive" users, that we should provide a link to a github PR tutorial at this point.

I'm happy to do so, if that matches your vision.

**Comment by charlieok** [comment]
I can see that you share my interest in writing good instructions, and paying attention to the details, and that's awesome. I remember thinking about some of these same issues when I wrote this:
- Can we give the user a command to copy and paste?
- Do we need to tell the user this? Most technical people can be assumed to know git, right?

The balance I struck was the note that's there now: “If there's a possibility you'll want to make and contribute changes, consider forking the repository and cloning from your fork.”

I wrote a variant of these instructions in an internal project, and struck a different balance there: it just tells them a command to run to clone from their fork, on the theory that if they're on the team, then of course they should just make their own fork.

I've also noticed that when I take some sparse docs and turn them into what I think are pretty good ones, the reader's standards go up accordingly. Small things that they still need to think about become issues to propose updates on. And that's actually great -- according to a sort of “broken window theory” (apologies for the unfortunate baggage behind that term), it suggests a raised quality bar is going to have some staying power.

On questions like this, I've come around to thinking that at some point you have to punt. You can't make the docs perfect. I run into this with differences between platforms too -- do you maintain a whole separate page when just a couple details are different, or add unsightly side notes to cover the differences?

I've landed on mostly accepting these kinds of small changes, even if it's not how I prefer it, if somebody is advocating for it and wants to do it. It lets someone scratch their itch, and continue along a path of more contributions, and maybe they are not alone and their fix really will make it easier for others coming after them.

So, given all that, how do you think this step should read?

**Comment by garethtdavies** [comment]
I like @charlieok approach here. I think it's safe to assume that anyone contributing to this repo would be comfortable with git (not least as the process itself is largely dependant on it). Adding a link to an external resource would be the best option in my opinion. 

The developer guide here is good for zcashd https://gitlab.com/zcash-docs/zcash-docs/blob/master/CONTRIBUTING.md and other than the URL are largely generic. I'm sure there a billion links as well we could link to for cloning repos from GitHub/Gitlab/Bitbucket et al.

> Good point! I usually clone from the "upstream" repo.. the one I want to contribute to, and then, locally add my fork as something-other-than-origin.

I'm the opposite 😄  I clone from my own fork and then add an "upstream" remote.

**Comment by zancas** [comment]
I'm not too concerned with which local alias points at which remote ('though it's fun to hear how @garethtdavies rolls).

When I suggested the change I was concerned with which authentication procedure is used.   I think the authentication mechanism offered by commands that start this way:
`git clone git@github.com:FOO`
is better than ones that start this way:
`git clone https://github.com/FOO`

   I think one uses PKI, where the other uses SSH.
   Of course, the SSH-based  (first) protocol is (at least in my experience) always bootstrapped off of `https`, but I don't think that renders it value-less to use SSH.

Given that another part of the instructions emphasizes setting up the SSH-based system, I think there's probably some agreement in the community.

In short.   I did a poor job of emphasizing what I thought was important, in the change that I suggested.    Sorry 'bout that.   I think the important thing to add to this doc is a clear directive to use git over `ssh`, instead of over `https`.   The change I suggested did that (and was copy-pasteable!) unfortunately it also necessarily includes a reference to a particular fork.

It's also possible that git will opt for ssh even if the alias points to a URL that starts with `https`, but I'd be a little surprised if it did that (at least by default).

**Comment by zancas** [comment]
I'm in transit, will propose new change when I get home.

**Comment by charlieok** [comment]
Your point that github ssh keys are likely wanted later (for writing signatures to the https://github.com/zcash/gitian.sigs/ project) is a good one. Even if a user isn't likely to submit any PRs to zcash-gitian, they are pretty likely to submit one to gitian.sigs.

I wasn't thinking of that when I wrote the step we're discussing here. That probably does tip the balance in favor of just preferring an ssh url by default. Thanks for raising that!

**Comment by zancas** [comment]
OK, I pushed a new version, what do you think?

**Comment by charlieok** [comment]
If you look at rest of the same README, you'll see that it uses an imaginary user with the github username `harrypotter`. I don't have any attachment to that but I want to try and be consistent.

**Comment by zancas** [comment]
Acknowledged.  I switched to `hpotter`, and added a note about permission bits on the `.ssh/config` file.

**Comment by charlieok** [comment]
almost there now... :)

The local username and github username were made deliberately different in order to remove some possible ambiguity. The section showing an example ssh config file entry illustrates that:

```
Host github.com
  User harrypotter
  PreferredAuthentications publickey
  IdentityFile /home/hpotter/.ssh/github_id_rsa
  AddKeysToAgent yes
```

**Comment by zancas** [comment]
Clever disambiguation.   Error fixed.

**Comment by charlieok** [comment]
Thanks!
